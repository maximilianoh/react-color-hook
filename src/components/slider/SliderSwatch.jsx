import React from 'react';
import reactCSS from 'reactcss';
import PropTypes from 'prop-types';

const SliderSwatch = ({
  hsl, offset, onClick, active, first, last,
}) => {
  const styles = reactCSS({
    default: {
      swatch: {
        height: '12px',
        background: `hsl(${hsl.h}, 50%, ${(offset * 100)}%)`,
        cursor: 'pointer',
      },
    },
    first: {
      swatch: {
        borderRadius: '2px 0 0 2px',
      },
    },
    last: {
      swatch: {
        borderRadius: '0 2px 2px 0',
      },
    },
    active: {
      swatch: {
        transform: 'scaleY(1.8)',
        borderRadius: '3.6px/2px',
      },
    },
  }, { active, first, last });

  const handleClick = (e) => onClick({
    h: hsl.h,
    s: 0.5,
    l: offset,
    source: 'hsl',
  }, e);

  return (
    <div style={styles.swatch} onKeyPress={handleClick} onClick={handleClick} role="grid" tabIndex={0} />
  );
};

SliderSwatch.defaultProps = {
  onClick: () => {},
  last: false,
  first: false,
  offset: 0,
  active: false,
};

SliderSwatch.propTypes = {
  onClick: PropTypes.func,
  hsl: PropTypes.shape({
    h: PropTypes.number,
  }).isRequired,
  offset: PropTypes.number,
  active: PropTypes.bool,
  last: PropTypes.bool,
  first: PropTypes.bool,
};

export default SliderSwatch;
