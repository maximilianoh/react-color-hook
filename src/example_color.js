export const red = {
  hsl: {
    a: 1, h: 0, l: 0.5, s: 1,
  },
  hex: '#ff0000',
  rgb: {
    r: 255, g: 0, b: 0, a: 1,
  },
  hsv: {
    h: 0, s: 1, v: 1, a: 1,
  },
};

export const yel = {
  hsl: {
    a: 1, h: 50, l: 0.71, s: 0.45,
  },
  hex: '#C4A921',
  rgb: {
    r: 196, g: 169, b: 33, a: 1,
  },
  hsv: {
    h: 50, s: 0.832, v: 0.769, a: 1,
  },
};

export const br = {
  hsl: {
    a: 1, h: 50, l: 0.30, s: 0.45,
  },
  hex: '#6F632A',
  rgb: {
    r: 111, g: 99, b: 42, a: 1,
  },
  hsv: {
    h: 50, s: 0.622, v: 0.435, a: 1,
  },
};
