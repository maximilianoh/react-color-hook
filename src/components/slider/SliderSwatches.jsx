import React from 'react';
import reactCSS from 'reactcss';
import PropTypes from 'prop-types';
import SliderSwatch from './SliderSwatch';

const SliderSwatches = ({ onClick, hsl }) => {
  const styles = reactCSS({
    default: {
      swatches: {
        marginTop: '20px',
      },
      swatch: {
        boxSizing: 'border-box',
        width: '20%',
        paddingRight: '1px',
        float: 'left',
      },
      clear: {
        clear: 'both',
      },
    },
  });

  // Acceptible difference in floating point equality
  const epsilon = 0.1;

  return (
    <div style={styles.swatches}>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.80}
          active={Math.abs(hsl.l - 0.80) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon}
          onClick={onClick}
          first
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.65}
          active={Math.abs(hsl.l - 0.65) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon}
          onClick={onClick}
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.50}
          active={Math.abs(hsl.l - 0.50) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon}
          onClick={onClick}
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.35}
          active={Math.abs(hsl.l - 0.35) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon}
          onClick={onClick}
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.20}
          active={Math.abs(hsl.l - 0.20) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon}
          onClick={onClick}
          last
        />
      </div>
      <div style={styles.clear} />
    </div>
  );
};

SliderSwatches.propTypes = {
  onClick: PropTypes.func,
  hsl: PropTypes.shape({
    l: PropTypes.number,
    s: PropTypes.number,
  }).isRequired,
};

SliderSwatches.defaultProps = {
  onClick: () => {},
};

export default SliderSwatches;
