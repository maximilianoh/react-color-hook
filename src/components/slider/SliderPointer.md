SliderPointer
```js
import SliderPointer from './SliderPointer.jsx';
import { red } from '../../example_color';
<SliderPointer />
```