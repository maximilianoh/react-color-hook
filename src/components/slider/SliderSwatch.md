SliderSwatch
```js
import SliderSwatch from './SliderSwatch.jsx';
import { red } from '../../example_color';
<SliderSwatch { ...red } />
```