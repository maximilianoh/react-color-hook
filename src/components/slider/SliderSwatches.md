SliderSwatches
```js
import SliderSwatches from './SliderSwatches.jsx';
import { red } from '../../helpers/color';
<SliderSwatches { ...red } />
```