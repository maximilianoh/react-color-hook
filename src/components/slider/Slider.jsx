import React from 'react';
import PropTypes from 'prop-types';
import reactCSS from 'reactcss';
import merge from 'lodash/merge';
import ColorWrap from '../common/ColorWrap';
import Hue from '../common/Hue';
import SliderSwatches from './SliderSwatches';
import SliderPointer from './SliderPointer';

const Slider = ({
  hsl, onChange, pointer,
  styles: passedStyles, className,
}) => {
  const styles = reactCSS(merge({
    default: {
      hue: {
        height: '12px',
        position: 'relative',
      },
      Hue: {
        radius: '2px',
      },
    },
  }, passedStyles));
  return (
    <div style={styles.wrap || {}} className={`slider-picker ${className}`}>
      <div style={styles.hue}>
        <Hue
          style={styles.Hue}
          hsl={hsl}
          pointer={pointer}
          onChange={onChange}
        />
      </div>
      <div style={styles.swatches}>
        <SliderSwatches hsl={hsl} onClick={onChange} />
      </div>
    </div>
  );
};

Slider.propTypes = {
  styles: PropTypes.shape({}),
  pointer: PropTypes.func,
  hsl: PropTypes.shape({}).isRequired,
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string,
};
Slider.defaultProps = {
  pointer: SliderPointer,
  styles: {},
  className: '',
};

export default ColorWrap(Slider);
