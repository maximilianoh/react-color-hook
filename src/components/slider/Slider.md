Slider
```js
import Slider from './Slider.jsx';
import { red } from '../../example_color';
<Slider { ...red } />
```


Slider custom style
```js
import Slider from './Slider.jsx';
<Slider styles={{ default: { wrap: { boxShadow: "0 1px 4px 0 rgba(200, 0, 0, 0.8)" } } }} />
```