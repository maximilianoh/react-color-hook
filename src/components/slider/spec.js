import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react';
import { red, yel, br } from '../../example_color';
import Slider from './Slider'
import SliderPointer from './SliderPointer'
import SliderSwatch from './SliderSwatch'
import SliderSwatches from './SliderSwatches'

test('Slider renders correctly', () => {
  const tree = renderer.create(
    <Slider { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Slider renders custom styles correctly', () => {
  const tree = renderer.create(
    <Slider styles={{ default: { wrap: { boxShadow: 'none' } } }} />,
  ).toJSON()
  expect(tree.props.style.boxShadow).toBe('none')
})


test('SliderPointer renders correctly', () => {
  const tree = renderer.create(
    <SliderPointer />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SliderSwatch renders correctly', () => {
  const tree = renderer.create(
    <SliderSwatch { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SliderSwatches renders correctly', () => {
  const tree = renderer.create(
    <SliderSwatches { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
});


test('SliderSwatch empty event', () => {
  render(<SliderSwatch { ...red } />);
  const d = document.querySelector("div");
  fireEvent.click(d.children[0]);
})


test('SliderSwatches empty event', () => {
  render(<SliderSwatches { ...red } />);
  const d = document.querySelector("div div");
  expect(d.children.length).toBe(6);
  fireEvent.click(d.children[0].children[0]);
});

test('SliderSwatches color epsilon', () => {
  render(<SliderSwatches { ...yel } />);
  const d = document.querySelector("div div");
  expect(d.children.length).toBe(6);
  fireEvent.click(d.children[0].children[0]);
});

test('SliderSwatches color epsilon 2', () => {
  render(<SliderSwatches { ...br } />);
  const d = document.querySelector("div div");
  expect(d.children.length).toBe(6);
  fireEvent.click(d.children[0].children[0]);
});