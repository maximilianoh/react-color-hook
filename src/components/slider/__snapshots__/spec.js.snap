// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders correctly 1`] = `
<div
  className="slider-picker "
  style={Object {}}
>
  <div
    style={
      Object {
        "height": "12px",
        "position": "relative",
      }
    }
  >
    <div
      style={
        Object {
          "MozBorderRadius": "",
          "MozBoxShadow": "",
          "OBorderRadius": "",
          "OBoxShadow": "",
          "WebkitBorderRadius": "",
          "WebkitBoxShadow": "",
          "borderRadius": "",
          "bottom": "0px",
          "boxShadow": "",
          "left": "0px",
          "msBorderRadius": "",
          "msBoxShadow": "",
          "position": "absolute",
          "right": "0px",
          "top": "0px",
        }
      }
    >
      <div
        className="hue-horizontal"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        style={
          Object {
            "MozBorderRadius": "",
            "OBorderRadius": "",
            "WebkitBorderRadius": "",
            "borderRadius": "",
            "height": "100%",
            "msBorderRadius": "",
            "outline": "none",
            "padding": "0 2px",
            "position": "relative",
          }
        }
        tabIndex={0}
      >
        <style>
          
          .hue-horizontal {
            background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
              33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            background: -webkit-linear-gradient(to right, #f00 0%, #ff0
              17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
          }

          .hue-vertical {
            background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
              #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
              #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
          }
        
        </style>
        <div
          style={
            Object {
              "left": "69.55128205128206%",
              "position": "absolute",
            }
          }
        >
          <div
            style={
              Object {
                "MozBorderRadius": "6px",
                "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "MozTransform": "translate(-7px, -1px)",
                "OBorderRadius": "6px",
                "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "OTransform": "translate(-7px, -1px)",
                "WebkitBorderRadius": "6px",
                "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "WebkitTransform": "translate(-7px, -1px)",
                "backgroundColor": "rgb(248, 248, 248)",
                "borderRadius": "6px",
                "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "height": "14px",
                "msBorderRadius": "6px",
                "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "msTransform": "translate(-7px, -1px)",
                "transform": "translate(-7px, -1px)",
                "width": "14px",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "marginTop": "20px",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="grid"
          style={
            Object {
              "MozBorderRadius": "2px 0 0 2px",
              "OBorderRadius": "2px 0 0 2px",
              "WebkitBorderRadius": "2px 0 0 2px",
              "background": "hsl(250.38461538461542, 50%, 80%)",
              "borderRadius": "2px 0 0 2px",
              "cursor": "pointer",
              "height": "12px",
              "msBorderRadius": "2px 0 0 2px",
            }
          }
          tabIndex={0}
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="grid"
          style={
            Object {
              "background": "hsl(250.38461538461542, 50%, 65%)",
              "cursor": "pointer",
              "height": "12px",
            }
          }
          tabIndex={0}
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="grid"
          style={
            Object {
              "background": "hsl(250.38461538461542, 50%, 50%)",
              "cursor": "pointer",
              "height": "12px",
            }
          }
          tabIndex={0}
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="grid"
          style={
            Object {
              "background": "hsl(250.38461538461542, 50%, 35%)",
              "cursor": "pointer",
              "height": "12px",
            }
          }
          tabIndex={0}
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          onKeyPress={[Function]}
          role="grid"
          style={
            Object {
              "MozBorderRadius": "0 2px 2px 0",
              "MozTransform": "scaleY(1.8)",
              "OBorderRadius": "0 2px 2px 0",
              "OTransform": "scaleY(1.8)",
              "WebkitBorderRadius": "0 2px 2px 0",
              "WebkitTransform": "scaleY(1.8)",
              "background": "hsl(250.38461538461542, 50%, 20%)",
              "borderRadius": "0 2px 2px 0",
              "cursor": "pointer",
              "height": "12px",
              "msBorderRadius": "0 2px 2px 0",
              "msTransform": "scaleY(1.8)",
              "transform": "scaleY(1.8)",
            }
          }
          tabIndex={0}
        />
      </div>
      <div
        style={
          Object {
            "clear": "both",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`SliderPointer renders correctly 1`] = `
<div
  style={
    Object {
      "MozBorderRadius": "6px",
      "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "MozTransform": "translate(-7px, -1px)",
      "OBorderRadius": "6px",
      "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "OTransform": "translate(-7px, -1px)",
      "WebkitBorderRadius": "6px",
      "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "WebkitTransform": "translate(-7px, -1px)",
      "backgroundColor": "rgb(248, 248, 248)",
      "borderRadius": "6px",
      "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "height": "14px",
      "msBorderRadius": "6px",
      "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "msTransform": "translate(-7px, -1px)",
      "transform": "translate(-7px, -1px)",
      "width": "14px",
    }
  }
/>
`;

exports[`SliderSwatch renders correctly 1`] = `
<div
  onClick={[Function]}
  onKeyPress={[Function]}
  role="grid"
  style={
    Object {
      "background": "hsl(0, 50%, 0%)",
      "cursor": "pointer",
      "height": "12px",
    }
  }
  tabIndex={0}
/>
`;

exports[`SliderSwatches renders correctly 1`] = `
<div
  style={
    Object {
      "marginTop": "20px",
    }
  }
>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      onKeyPress={[Function]}
      role="grid"
      style={
        Object {
          "MozBorderRadius": "2px 0 0 2px",
          "OBorderRadius": "2px 0 0 2px",
          "WebkitBorderRadius": "2px 0 0 2px",
          "background": "hsl(0, 50%, 80%)",
          "borderRadius": "2px 0 0 2px",
          "cursor": "pointer",
          "height": "12px",
          "msBorderRadius": "2px 0 0 2px",
        }
      }
      tabIndex={0}
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      onKeyPress={[Function]}
      role="grid"
      style={
        Object {
          "background": "hsl(0, 50%, 65%)",
          "cursor": "pointer",
          "height": "12px",
        }
      }
      tabIndex={0}
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      onKeyPress={[Function]}
      role="grid"
      style={
        Object {
          "background": "hsl(0, 50%, 50%)",
          "cursor": "pointer",
          "height": "12px",
        }
      }
      tabIndex={0}
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      onKeyPress={[Function]}
      role="grid"
      style={
        Object {
          "background": "hsl(0, 50%, 35%)",
          "cursor": "pointer",
          "height": "12px",
        }
      }
      tabIndex={0}
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      onKeyPress={[Function]}
      role="grid"
      style={
        Object {
          "MozBorderRadius": "0 2px 2px 0",
          "OBorderRadius": "0 2px 2px 0",
          "WebkitBorderRadius": "0 2px 2px 0",
          "background": "hsl(0, 50%, 20%)",
          "borderRadius": "0 2px 2px 0",
          "cursor": "pointer",
          "height": "12px",
          "msBorderRadius": "0 2px 2px 0",
        }
      }
      tabIndex={0}
    />
  </div>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;
